contrib: Fixup verify-binaries OS platform parsing

BenWestgate opened the pull request:
Closes #30145.

This PR solves two major issues in the `parse_version_string` function of verify-binaries:
1. `-aarch64` binaries cannot be specifically downloaded. The -platform string gets interpreted as a release candidate that doesn't exist due to containing sub-string "rc".
2. Specifying a platform with a "-" in the name causes the parser to ignore both "-platform" AND "-rcN" and download the potentially wrong (non-rc) version for every platform. This also prevented specifying just one platform binary the user wished to download.

It also updates the accompanying `test.py` to cover problem two and adds two examples that were formerly broken to `README.md` to show what is now possible. Including the most useful case of downloading only 1 specific platform's binary.

This improves the Bitcoin verify-binaries tools user experience by not:
1. Failing to download for inexplicable reasons,
2. Downloading more files than what the user told it to, or in the worst case
3. Downloading only the wrong files.

* A test was added to cover the command `verify-binaries/verify.py pub 22.0-x86_64-linux-gnu.tar.gz` which checks that _bitcoin-22.0-x86_64-linux-gnu.tar.gz_ downloads successfully AND ONLY _bitcoin-22.0-x86_64-linux-gnu.tar.gz_ downloads.
* The steps to reproduce each bug are in the referenced issue #30145. Explanation of the potential issue as well as reasoning for the way the bug was fixed are in my commit descriptions.
* This delivers the promised feature of "only download the binaries for a certain platform", by allowing strings with '-' to be accepted, allowing for single file downloads for any specific platform which was not always possible before.
* Removes 6 lines of code from the offending `parse_version_string` function, while fixing the bugs/errors, and extending the functionality to be practical for users with slow connections.
* Makes the error message more helpful when no file matches the provided platform string, now prints "Did you mean: `closest-match`" to help correct typos.

Thanks for reading my PR. I look forward to getting this helpful tool in its best shape yet.

Log of this branch passing the new test.py:
```
python3 test.py
✓ 'Nonexistent version should fail' passed
✓ 'Malformed version should fail' passed
✓ '--min-good-sigs 20 should fail' passed
- testing verification (22.0-x86_64-linux-gnu.tar.gz)
✓ '22.0-x86_64-linux-gnu.tar.gz should succeed' passed
- testing verification (22.0)
✓ '22.0 should succeed' passed
```

Log of master failing the new test.py:
```
python3 test.py
✓ 'Nonexistent version should fail' passed
✓ 'Malformed version should fail' passed
✓ '--min-good-sigs 20 should fail' passed
- testing verification (22.0-x86_64-linux-gnu.tar.gz)
✓ '22.0-x86_64-linux-gnu.tar.gz should succeed' passed
Traceback (most recent call last):
  File "/home/ben/Documents/GitHub/bitcoin/contrib/verify-binaries/test.py", line 74, in <module>
    main()
  File "/home/ben/Documents/GitHub/bitcoin/contrib/verify-binaries/test.py", line 27, in main
    assert len(v) == 1
           ^^^^^^^^^^^
AssertionError
```

BenWestgate commented:
@DrahtBot: I am fairly sure this PR is NOT a consensus change. It's a bugfix to a download verification tool script. Replace that label with a more accurate one.

willcl-ark commented:
There are also similar changes in https://github.com/bitcoin/bitcoin/pull/28418 but I didn't check this PR yet to see if they are the same.

BenWestgate commented:
> There are also similar changes in #28418 but I didn't check this PR yet to see if they are the same.

This PR includes a **test.py** for the new functionality. My diff is smaller: Only 3 adds to `verify.py` 
Either could close my linked issue.

I like my parser better and #28418 's error messages. I will borrow his strings and squash this.

Also my bad for almost duplicating someone else's PR. I did search for "verify.py" but didn't find any issues. I'll give #28418 a review.

BenWestgate committed:
contrib: Fixup verify-binaries OS platform parsing

Parse platform strings with "-" or '.' correctly such as "linux-gnu" or
"x86_64-linux-gnu.tar.gz" to download the matching files or file. String
partition() is used to tolerate more dashes. Update `VERSION_EXAMPLE`
with a new string parsed correctly now. Fix "-aarch64" interpreted as a
release candidate due to sub-string "rc", causing all downloads to fail.
Now "rc" must immediately follow first "-" to indicate an [-rc] string.
Local variables `version_rc`, `version_os` renamed to `rc`, `platform`.
If "-rcN" is specified, `platform` is reassigned to remove the '-rcN'.

Changes are useful to only download one bitcoin core binary on slow
connections. Making `verify.py pub` more intuitive, robust, and
versatile. Closes #30145

When user types a platform string not found in any filename lets help
and say the platform closest to what they typed in a `f"No files
matched the platform specified. Did you mean: {closest_match}"` log.
Improves UX when unaware how we name our files.
Uses the difflib Python built-in which was already imported elsewhere.

Update test.py to test single file verification
verify-binaries/verify.py can accept an entire filename filter for its
"-platform" parameter now so let us test that it succeeds and downloads
and verifies only one file. `verify.py pub 22.0-x86_64-linux-gnu.tar.gz`
should get and verify only the requested binary. It is placed before the
existing <version> wide verification as it is a faster test and possibly
easier to break.

Update doc with examples now possible after bugfix
Add example to show release candidates now work with "-platform" strings
containing "-" and string provided can be from the middle of filename:
`./contrib/verify-binaries/verify.py --json pub 23.0-rc5-linux-gnu`
Change example 5 to not match example 3.
New examples to show platform can now be provided specifically enough to
download only a single binary down to its file extension:
`./contrib/verify-binaries/verify.py pub 25.2-x86_64-linux`
`./contrib/verify-binaries/verify.py pub 24.1-rc1-darwin`
`./contrib/verify-binaries/verify.py pub 27.0-win64-setup.exe`
This is the most common use if not verifying all files so users see it
as the first example for "only download the binaries for a certain
architecture and/or platform". Downloading one file is intuitively what
most will think this meant and this change delivers on that expectation.

Co-authored-by: stickies-v

BenWestgate commented:
> Merge commits in pull requests are discouraged in this repo, please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

Done.

willcl-ark commented:
Yep I agree this is cleaner and works nicely in my initial manual testing :)

I will do a code review early next week, thanks!

willcl-ark reviewed (commented):
I think the version parser here is is pretty good shape generally, or at least improved from its current state.

I am not convinced yet that it is handling `rc` correctly yet though, see results of this additional test i added:

```log
₿ ./contrib/verify-binaries/test2.py
Checking 27.0
Found RC version: version_base=27.0, version_rc= os_filter=
Checking 27.0-x86_64
Found RC version: version_base=27.0, version_rc= os_filter=x86_64
Checking 27.0-linux
Found RC version: version_base=27.0, version_rc= os_filter=linux
Checking 27.0-win64.zip
Found RC version: version_base=27.0, version_rc= os_filter=win64.zip
Checking bitcoin-27.0.tar.gz
Found RC version: version_base=bitcoin, version_rc= os_filter=27.0.tar.gz
Checking bitcoin-27.1-rc1.tar.gz
Found RC version: version_base=bitcoin, version_rc= os_filter=27.1-rc1.tar.gz
Missing rc
Checking 23.0-rc5-linux-gnu
Found RC version: version_base=23.0, version_rc=rc5 os_filter=linux-gnu
Checking bitcoin-27.0-rc1-x86_64-linux-gnu
Found RC version: version_base=bitcoin, version_rc= os_filter=27.0-rc1-x86_64-linux-gnu
Missing rc
Checking bitcoin-27.0-rc1-x86_64-linux-gnu.tar.gz
Found RC version: version_base=bitcoin, version_rc= os_filter=27.0-rc1-x86_64-linux-gnu.tar.gz
Missing rc
```

Passing an architecture and platform, or an extension sees it miss-parse `version_rc`.

Without a live release candidate build being hosted anywhere, I am not sure what exactly `bitcoin-27.0-rc1-x86_64-linux-gnu` would download with this `os_filter`. I could probably try and work it out, but didn't bother here yet.

willcl-ark commented:
Here is the file I was using to check the new parsing

<details>
<summary>Details</summary>

```python
#!/usr/bin/env python3

import tempfile
from verify import parse_version_string, parse_sums_file

SHASUMS = """cb35e250ae9d0328aa90e7aad0b877ed692597420a1092e8ab1a5dd756209722  bitcoin-27.0-aarch64-linux-gnu.tar.gz
9d4c28e7620d03bf346ebea388f222e4d6d2b996d7eb32fab72707b8320d5249  bitcoin-27.0-arm-linux-gnueabihf.tar.gz
7f060f2cd07746ff9d09b000b4195fee88dfca8444ab7a73f0c76aff4225227c  bitcoin-27.0-arm64-apple-darwin.zip
1d9d9b837297a73fc7a3b1cfed376644e3fa25c4e1672fbc143d5946cb52431d  bitcoin-27.0-arm64-apple-darwin.tar.gz
9c1ee651d3b157baccc3388be28b8cf3bfcefcd2493b943725ad6040ca6b146b  bitcoin-27.0.tar.gz
6ceaedb59ca33b751387b15f2c8da7f2f7cd2739c6464fc6cbef440852869b92  bitcoin-27.0-powerpc64-linux-gnu.tar.gz
3c00f81a7c67b4cf3e382fae7eaa2c7facea2dfdf39f4c281512237c06b71960  bitcoin-27.0-powerpc64le-linux-gnu.tar.gz
371e53b21c3ba29a90e69c30b7213d75c165d084bde50ae6d73ee0e1ef179e68  bitcoin-27.0-riscv64-linux-gnu.tar.gz
8c94d3a7e34b59effdcf283263d5e84f2b009e601076282e9697ab4244bef3e8  bitcoin-27.0-x86_64-apple-darwin.zip
e1efd8c4605b2aabc876da93b6eee2bedd868ce7d1f02b0220c1001f903b3e2c  bitcoin-27.0-x86_64-apple-darwin.tar.gz
2a6974c5486f528793c79d42694b5987401e4a43c97f62b1383abf35bcee44a8  bitcoin-27.0-x86_64-linux-gnu.tar.gz
a2aa3db390a768383e8556878250a44f3eb3b7a6e91e94e47fa35c06b6e8d09f  bitcoin-27.0-win64-setup.exe
ca75babeaa3fb75f5a166f544adaa93fd7c1f06cf20d4e2c8c2a8b010f4c7603  bitcoin-27.0-win64.zip
"""

expected_results = [
    (
        "27.0",
        [
            "bitcoin-27.0-aarch64-linux-gnu.tar.gz",
            "bitcoin-27.0-arm-linux-gnueabihf.tar.gz",
            "bitcoin-27.0-arm64-apple-darwin.zip",
            "bitcoin-27.0-arm64-apple-darwin.tar.gz",
            "bitcoin-27.0.tar.gz",
            "bitcoin-27.0-powerpc64-linux-gnu.tar.gz",
            "bitcoin-27.0-powerpc64le-linux-gnu.tar.gz",
            "bitcoin-27.0-riscv64-linux-gnu.tar.gz",
            "bitcoin-27.0-x86_64-apple-darwin.zip",
            "bitcoin-27.0-x86_64-apple-darwin.tar.gz",
            "bitcoin-27.0-x86_64-linux-gnu.tar.gz",
            "bitcoin-27.0-win64-setup.exe",
            "bitcoin-27.0-win64.zip",
        ],
    ),
    (
        "27.0-x86_64",
        [
            "bitcoin-27.0-x86_64-apple-darwin.zip",
            "bitcoin-27.0-x86_64-apple-darwin.tar.gz",
            "bitcoin-27.0-x86_64-linux-gnu.tar.gz",
        ],
    ),
    (
        "27.0-linux",
        [
            "bitcoin-27.0-aarch64-linux-gnu.tar.gz",
            "bitcoin-27.0-arm-linux-gnueabihf.tar.gz",
            "bitcoin-27.0-powerpc64-linux-gnu.tar.gz",
            "bitcoin-27.0-powerpc64le-linux-gnu.tar.gz",
            "bitcoin-27.0-riscv64-linux-gnu.tar.gz",
            "bitcoin-27.0-x86_64-linux-gnu.tar.gz",
        ],
    ),
    ("27.0-win64.zip", ["bitcoin-27.0-win64.zip"]),
    ("bitcoin-27.0.tar.gz", ["bitcoin-27.0.tar.gz"]),
    ("bitcoin-27.1-rc1.tar.gz", ["bitcoin-27.0.tar.gz"]),
    ("23.0-rc5-linux-gnu", []),
    ("bitcoin-27.0-rc1-x86_64-linux-gnu", []),
    ("bitcoin-27.0-rc1-x86_64-linux-gnu.tar.gz", []),
]


def test_version_parser():
    with tempfile.NamedTemporaryFile(mode="w+t", delete=False) as tmp_sums_file:
        tmp_sums_file.write(SHASUMS)
        tmp_sums_file_path = tmp_sums_file.name

    for term, matches in expected_results:
        print(f"Checking {term}")
        version_base, version_rc, os_filter = parse_version_string(term)
        print(f"Found RC version: {version_base=:}, {version_rc=:} {os_filter=:}")
        if not version_rc and ("rc" in term):
            print("Missing rc")
        hashes_to_verify = parse_sums_file(tmp_sums_file_path, [os_filter])
        results = [name for hash, name in hashes_to_verify]
        try:
            assert sorted(matches) == sorted(results)
        except:
            continue


if __name__ == "__main__":
    test_version_parser()
```

</details>

BenWestgate commented:
Here’s the issue:

> Checking bitcoin-27.0.tar.gz
> Found RC version: version_base=bitcoin, version_rc= os_filter=27.0.tar.gz

It’s not designed to work when the version string begins with bitcoin-
What else would we be using bitcoin/bitcoin contrib/verify-binaries for?

27.0-.tar.gz would be the correct way to search for all binaries ending in this extension with this parser.

Whatever comes before the 1st dash must be the version number in <major>.<minor> or <major>.<minor>.<patch> format.

Even without starting with bitcoin-
27.0.tar.gz should be rejected as it’s not a valid version number and the first field before ‘-‘ is always the version number.

Do you want it to not get confused when the string starts with bitcoin before the version number?

> Not sure if we want this to specify bitcoincore.org,bitcoin.orgdoes host some binaries too (and recently even added a few which aren't totally out of date!), and we do try both hosts. On the other hand, this advice is actually more useful, accurate and likely to be correct, so 🤷🏼‍♂️

For this, I thought the script should actually print the available versions that it did find rather than tell the user to navigate to a website.

That is a little out of the scope of this PR but I’d be happy to update that warning to spit out all the file names (or search strings that would exact match a file name, i.e. with “bitcoin-“ removed.) or better: the closest matches to the platform they searched for by edit distance.

That way something like 27.0-win65.zop would suggest “27.0-win64.zip” first in the list of “did you mean?” Instead of printing every platform.

It looks like it’s parsing -rcN correctly when it’s the 2nd field after version number.

On Wed, May 29, 2024 at 05:32, Will Clark ***@***.***(mailto:On Wed, May 29, 2024 at 05:32, Will Clark <<a href=)> wrote:

> @willcl-ark commented on this pull request.
>
> I think the version parser here is is pretty good shape generally, or at least improved from its current state.
>
> I am not convinced yet that it is handling rc correctly yet though, see results of this additional test i added:
>
> ₿ ./contrib/verify-binaries/test2.py
> Checking 27.0
> Found RC version: version_base=27.0, version_rc= os_filter=
> Checking 27.0-x86_64
> Found RC version: version_base=27.0, version_rc= os_filter=x86_64
> Checking 27.0-linux
> Found RC version: version_base=27.0, version_rc= os_filter=linux
> Checking 27.0-win64.zip
> Found RC version: version_base=27.0, version_rc= os_filter=win64.zip
> Checking bitcoin-27.0.tar.gz
> Found RC version: version_base=bitcoin, version_rc= os_filter=27.0.tar.gz
> Checking bitcoin-27.1-rc1.tar.gz
> Found RC version: version_base=bitcoin, version_rc= os_filter=27.1-rc1.tar.gz
> Missing rc
> Checking 23.0-rc5-linux-gnu
> Found RC version: version_base=23.0, version_rc=rc5 os_filter=linux-gnu
> Checking bitcoin-27.0-rc1-x86_64-linux-gnu
> Found RC version: version_base=bitcoin, version_rc= os_filter=27.0-rc1-x86_64-linux-gnu
> Missing rc
> Checking bitcoin-27.0-rc1-x86_64-linux-gnu.tar.gz
> Found RC version: version_base=bitcoin, version_rc= os_filter=27.0-rc1-x86_64-linux-gnu.tar.gz
> Missing rc
>
> Passing an architecture and platform, or an extension sees it miss-parse version_rc.
>
> Without a live release candidate build being hosted anywhere, I am not sure what exactly bitcoin-27.0-rc1-x86_64-linux-gnu would download with this os_filter. I could probably try and work it out, but didn't bother here yet.
>
> ---------------------------------------------------------------
>
> In [contrib/verify-binaries/verify.py](https://github.com/bitcoin/bitcoin/pull/30147#discussion_r1618650022):
>
>> @@ -514,7 +509,7 @@ def cleanup():
>      # Extract hashes and filenames
>      hashes_to_verify = parse_sums_file(SUMS_FILENAME, [os_filter])
>      if not hashes_to_verify:
> -        log.error("no files matched the platform specified")
> +        log.error("no files matched the platform specified. check available versions on bitcoincore.org/bin")
>
> Not sure if we want this to specify bitcoincore.org, bitcoin.org does host some binaries too (and recently even added a few which aren't totally out of date!), and we do try both hosts. On the other hand, this advice is actually more useful, accurate and likely to be correct, so 🤷🏼‍♂️
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/bitcoin/bitcoin/pull/30147#pullrequestreview-2085013260), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ARQZ6FZ5TPH6HCQJ4XLCO73ZEWVEDAVCNFSM6AAAAABICBAWPSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAOBVGAYTGMRWGA).
> You are receiving this because you modified the open/close state.Message ID: ***@***.***>

willcl-ark reviewed (approved):
tACK 8135632c33f4bad8434403b5807c48d7dba3b3d7

This works much better than the current verifier, and being able to specify an exact version to download is more convenient in general, and especially in places such as [building docker images](https://github.com/willcl-ark/bitcoin-core-docker/blob/803e20df0f2f70ca6a635f378590e6220de743b3/27/Dockerfile#L25-L27) where downloading multiple unneeded versions is purely wasteful.

Still one open comment regarding specifying bitcoincore.org, but happy to reACK if that gets changed, and it's not a blocker for me.

BenWestgate commented on contrib/verify-binaries/verify.py:
```diff
@@ -514,7 +509,7 @@ def cleanup():
     hashes_to_verify = parse_sums_file(SUMS_FILENAME, [os_filter])
     if not hashes_to_verify:
-        log.error("no files matched the platform specified")
+        log.error("no files matched the platform specified. check available versions on bitcoincore.org/bin")
```
https://github.com/bitcoin/bitcoin/blob/5cc2ae4ff287c6fef10b2e634e35353588f9ea3d/contrib/verify-binaries/verify.py#L511-L515

willcl-ark replied:
Well, I've enjoyed learning that python has a `difflib` builtin library. I thought I'd used all the string libraries by now, but somehow never come across that one.

If you did wwant to incorporate that change here, then I'd be happy to reACK this. I did a few tests and it did for example nicely match `27.1-x64-linux` to suggest `Did you mean: x86_64-linux-gnu.tar.gz`, along with a few other typos I tried, which does seem neat.

BenWestgate replied:
Incorporated.

BenWestgate commented:
> tACK [8135632](https://github.com/bitcoin/bitcoin/commit/8135632c33f4bad8434403b5807c48d7dba3b3d7)
> 
> Still one open comment regarding specifying bitcoincore.org, but happy to reACK if that gets changed, and it's not a blocker for me.

I wrote a better error message that gives a "Did you mean: <closest_match>" when the platform typed isn't in the SHA256SUMS. Avoiding the domain issue and being significantly more helpful. LMK if you like it enough to add to this PR.

willcl-ark reviewed (approved):
ACK a2fc9ddee9cbaeffb3c460973bab3c2dd734db55

Compared changes since last review using `git range-diff 8135632...a2fc9ddee9cbaeffb3c460973bab3c2dd734db55`, which added the `difflib` helper and updated the commit message.

stickies-v commented on contrib/verify-binaries/verify.py:
```diff
@@ -514,7 +509,9 @@ def cleanup():
-        log.error("no files matched the platform specified")
+        available_versions = ["-".join(line[1].split("-")[2:]) for line in parse_sums_file(SUMS_FILENAME, [])]
+        closest_match = difflib.get_close_matches(os_filter, available_versions, cutoff=0, n=1)[0]
+        log.error("No files matched the platform specified. Did you mean: "+closest_match)
```
f-string nit
```suggestion
        log.error(f"No files matched the platform specified. Did you mean: {closest_match}")
```

stickies-v reviewed (approved):
ACK a2fc9ddee9cbaeffb3c460973bab3c2dd734db55

Nice improvement, a few suggestion nits but not blocking - happy to quickly re-review if you adopt them.

Tested that all the new examples work too.

BenWestgate commented on contrib/verify-binaries/verify.py:
```diff
@@ -97,21 +97,16 @@ def bool_from_env(key, default=False) -> bool:
-            version_os = parts[1]
-    elif len(parts) == 3:  # "<version>-rcN-platform"
+    version_os = "-".join(parts[1:]) # "<version>" or "<version>-platform"
+    if "-rc" in "-".join(parts[:2]): # "<version>-rcN[-platform]"
```
That does look nicer, implementing your suggestion. Thanks for teaching me `.partition()` and `.startswith()`.

stickies-v replied:
Great. I didn't mean to take out your example docstrings btw so feel free to add those/similar ones in, it's helpful.

BenWestgate replied:
```python3
def parse_version_string(version_str):
    parts = version_str.split('-')
    # "<version>" or "<version>-platform"
    version_base, _, version_os = version_str.partition('-')
    version_rc = ""
    if version_os.startswith("rc"): # "<version>-rcN[-platform]"
        version_rc, _, version_os = version_os.partition('-')

    return version_base, version_rc, version_os
```

Look good to you?

stickies-v replied:
Actually: would `version`, `rc`, `platform` be better variable names?

stickies-v replied:
> Look good to you?

No need for the `parts = ` statement I think.

> `# "<version>" or "<version>-platform"`

That seems wrong, given that we may also have an `rc` part still?

BenWestgate replied:
All of your comments are right! Glad I asked.

Looks better with the new local variable names:

```python3

def parse_version_string(version_str):
    # "<version>[-rcN][-platform]"
    version_base, _, platform = version_str.partition('-')
    rc = ""
    if platform.startswith("rc"): # "<version>-rcN[-platform]"
        rc, _, platform = platform.partition('-')
    # else "<version>" or "<version>-platform"

    return version_base, rc, platform
```

I don't love my first comment, but it's less clear without it.

stickies-v reviewed (approved):
ACK 4a85b154d0874f7bfa848ae8d7ab341eea754607 - nice work

BenWestgate commented on contrib/verify-binaries/verify.py:
```diff
@@ -468,7 +462,7 @@ def cleanup():
     # determine remote dir dependent on provided version string
     try:
-        version_base, version_rc, os_filter = parse_version_string(args.version)
+        version_base, rc, os_filter = parse_version_string(args.version)
```
It was a "change all occurrences" accident. `version_os` was local to `parse_version_string`, so I logically assumed `version_rc` would be too. So I blame the original variable names for this goof.
I didn't see until after I had force pushed.
 
Shall I restrict the `version_rc` to `rc` rename to only within the `parse_version_string` function? Or not worth changing back?

stickies-v replied:
I'd prefer limiting it to `parse_version_string` and will quickly re-ack if you force push

BenWestgate replied:
Done in https://github.com/bitcoin/bitcoin/compare/4a85b154d0874f7bfa848ae8d7ab341eea754607..3ab25201909bece9066ac6191670bcee09791d54

stickies-v reviewed (approved):
re-ACK 3ab25201909bece9066ac6191670bcee09791d54

willcl-ark reviewed (approved):
re-ACK 3ab25201909bece9066ac6191670bcee09791d54

Based on range diff since previous review: `git range-diff a2fc9ddee9cbaeffb3c460973bab3c2dd734db55...3ab25201909bece9066ac6191670bcee09791d54`
